zerver/lib/retention.py

```diff
@@ -458,35 +458,53 @@ def get_realms_and_streams_for_archiving() -> list[tuple[Realm, list[Stream]]]:
 def move_messages_to_archive(
     message_ids: list[int], realm: Realm | None = None, chunk_size: int = MESSAGE_BATCH_SIZE
 ) -> None:
-    # Uses index: zerver_message_pkey
-    query = SQL(
-        """
-    INSERT INTO zerver_archivedmessage ({dst_fields}, archive_transaction_id)
-        SELECT {src_fields}, {archive_transaction_id}
-        FROM zerver_message
-        WHERE zerver_message.id IN {message_ids}
-        LIMIT {chunk_size}
-    ON CONFLICT (id) DO UPDATE SET archive_transaction_id = {archive_transaction_id}
-    RETURNING id
     """
-    )
-    count = run_archiving_in_chunks(
-        query,
-        type=ArchiveTransaction.MANUAL,
-        message_ids=Literal(tuple(message_ids)),
-        realm=realm,
-        chunk_size=chunk_size,
-    )
+    Callers using this to archive a large amount of messages should ideally make sure the message_ids are
+    ordered, as that'll allow better performance here by keeping the batches that'll be sent to the database
+    ordered rather than randomly scattered.
+    """
+    count = 0
+    # In order to avoid sending a massive list of message ids to the database,
+    # we'll handle chunking the list of ids directly here.
+    message_ids_head = message_ids
+    while message_ids_head:
+        message_ids_chunk = message_ids_head[0:chunk_size]
+        message_ids_head = message_ids_head[chunk_size:]
+
+        # Uses index: zerver_message_pkey
+        query = SQL(
+            """
+        INSERT INTO zerver_archivedmessage ({dst_fields}, archive_transaction_id)
+            SELECT {src_fields}, {archive_transaction_id}
+            FROM zerver_message
+            WHERE zerver_message.id IN {message_ids}
+            LIMIT {chunk_size}
+        ON CONFLICT (id) DO UPDATE SET archive_transaction_id = {archive_transaction_id}
+        RETURNING id
+        """
+        )
+
+        # Since we already limited the message_ids to be used in the query to chunk_size,
+        # run_archiving_in_chunks will only have a single chunk to process, which might be
```

Comment:
I think the logic means it actually runs the query _twice_, since it gets exactly its chunk_size rows back, so it doesn't know if there are still more to pull.  So this doubles the number of queries, with 50% of them moving no rows

Reply:
Ohhh yeah, thanks for catching that - that completely fails with the `len(new_chunk) < chunk_size` loop terminating condition in `run_archiving_in_chunks` - repeats the query as you said.

Pushed with a fix now, eliminating the `LIMIT {chunk_size}` from the query as you suggested in the other comment; and adding support for `chunk_size=None` in `run_archiving` for exactly this kind of query, to not do looping. That solves this issue here without having to do anything hacky like passing `chunk_size+1` or something.

Comes with a prep commit to rename to just `run_archiving`, since this function won't always do chunking anymore (nor is it necessary to elaborate on that in the function name).